api/task_create.go

```diff
@@ -55,47 +55,60 @@ func (h *TaskLifeCycleHandler) CreateTask(w http.ResponseWriter, r *http.Request
 		return
 	}
 
+	// Initialize the new task
+	err = initNewTask(r.Context(), d)
+	if err != nil {
+		err = fmt.Errorf("error initializing new task: %s", err)
+		logger.Error("error creating task", "error", err)
+		sendError(w, r, http.StatusBadRequest, err.Error())
+		return
+	}
+
 	// Create a dry run task if run parameter is set to inspect
 	var run string
 	if params.Run != nil {
 		run = string(*params.Run)
 	}
 	if run == driver.RunOptionInspect {
+		logger.Trace("run inspect option", "task_name", d.Task().Name())
 		h.createDryRunTask(w, r, d, trc)
 		return
 	}
 
-	// Create a new event for tracking infrastructure change required actions
-	var storedErr error
-	task := d.Task()
-	ev, err := event.NewEvent(task.Name(), &event.Config{
-		Providers: task.ProviderNames(),
-		Services:  task.ServiceNames(),
-		Source:    task.Source(),
-	})
-	if err != nil {
-		err = fmt.Errorf("error creating new event: %s", err)
-		logger.Error("error creating task", "error", err)
-		sendError(w, r, http.StatusInternalServerError, err.Error())
-		return
-	}
-	defer func() {
-		ev.End(storedErr)
+	// Apply task if run option is now
+	if run == driver.RunOptionNow {
+		task := d.Task()
+		taskName := task.Name()
+		logger.Trace("run now option", "task_name", taskName)
+
+		// Create new event for task run
+		ev, err := event.NewEvent(taskName, &event.Config{
+			Providers: task.ProviderNames(),
+			Services:  task.ServiceNames(),
+			Source:    task.Source(),
+		})
+		if err != nil {
+			err = fmt.Errorf("error creating new event: %s", err)
+			logger.Error("error creating task", "error", err)
+			sendError(w, r, http.StatusInternalServerError, err.Error())
+			return
+		}
+		ev.Start()
+		// Apply task
+		err = d.ApplyTask(r.Context())
+		if err != nil {
+			err = fmt.Errorf("error applying new task: %s", err)
+			logger.Error("error applying task", "error", err)
+			sendError(w, r, http.StatusBadRequest, err.Error())
+			return
+		}
+		// Store event if apply was successful and task will be created
+		ev.End(err)
```

Comment:
It seems as though if the task is created but ApplyTask() errored, it would be useful to store an event to notify that the apply was unsuccessful. That is, as I understand, the task is still created (and not rolled back) and ran; it was just unsuccessful. I think I might be missing something. Let me know!

Reply:
In its current state, the task is rolled back if an apply errors. However, I've been thinking of changing the create and run request to be asynchronous since a task run could take an arbitrary amount of time. In this asynchronous workflow, I think we wouldn't do any roll back and then store the failed events so that the user has feedback on the run.

Reply:
Gotcha! Thanks for checking me. I see what you're saying - the new task's driver doesn't get added to the map of drivers until we know that the apply was successful. Sounds good